widget/embedded/src/lib/tabManager/tabManager.ts

```diff
@@ -0,0 +1,235 @@
+  Events,
+  ID,
+  Payload,
+  TabManagerInterface,
+} from './tabManager.types';
+
+import { CHANNEL_NAME, TIMEOUT } from './tabManager.constants';
+
+export class TabManager implements TabManagerInterface {
+  private id: ID;
+  private claimed: boolean | null = null;
+  private lastCheck: number | undefined;
+  private events: Events = {};
+  private channel: BroadcastChannel;
+  private tabIDs: ID[] = [];
+
+  constructor(events: Events) {
+    this.channel = new BroadcastChannel(CHANNEL_NAME);
+    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+    this.id = Math.trunc(Math.random() * 100000);
+    this.events = events;
+    this.initEvents();
+    this.tryClaim();
+  }
+
+  forceClaim = (): void => {
+    if (!this.claimed) {
+      const message: Payload = { name: 'force-claim', id: this.id };
+      this.channel.postMessage(message);
+    }
+  };
+
+  isClaimed() {
+    return !!this.claimed;
+  }
+
+  /**
+   *
+   * Attach some event listeners to the window.
+   * It includes listening to broadcast messages, entering and exiting the page event.
+   *
+   */
+  private initEvents() {
+    const message: Payload = { name: 'open', id: this.id };
+    this.channel.postMessage(message);
+
+    this.channel.addEventListener('message', (event: MessageEvent<Payload>) => {
+      const name = event.data.name;
+
+      switch (name) {
+        case 'ping':
+          this.claimedByCurrentTab(event.data.pingAt);
+          break;
+
+        case 'pong':
+          this.alreadyClaimedByAnotherTab();
+          break;
+
+        case 'release':
+          this.updateAndClaim(event.data.id, event.data.openedTabs);
+          break;
+
+        case 'open':
+          this.saveTabID(event.data.id);
+          break;
+
+        case 'close':
+          this.removeTabId(event.data.id);
+          break;
+
+        case 'force-claim':
+          this.forceRelease(event.data.id);
+          break;
+
+        case 'force-release':
+          this.updateAndClaim(event.data.id, event.data.openedTabs);
+          break;
+
+        default:
+          break;
+      }
+    });
+
+    window.addEventListener('beforeunload', () => {
+      if (this.claimed) {
+        this.release();
+      } else {
+        const message: Payload = { name: 'close', id: this.id };
+        this.channel.postMessage(message);
+      }
+    });
+  }
+  /**
+   *
+   * Runs after a `ping` message and checks if current tab is claimed the tabs,
+   *    if yes, it responds with `pong`, If not, it doesn't do anything.
+   *
+   * Note 1: If we receive a `ping` message,
+   *    it means a tab is trying to claim tabs directly, usually it's the active tab.
+   *
+   * Note 2: If multiple tabs are opening at once, all of them are broadcasting `ping`
+   *    and eventually all of them will `claim` the queue after the TIMEOUT.
+   *    For avoiding this, we will accepts the oldest broadcasted ping. (`isPingSentBefore`)
+   *
+   *
+   * @param {*} pingAt
+   * @returns
```

Comment:
I think we could remove this part as we didn't have similar jsdoc for others as well.